aws/client.go

```diff
@@ -0,0 +1,505 @@
+		EnsureSecurityGroupIngressRules(ctx context.Context, securityGroupID string) error
+		EnsureAndAssociateElasticIPv4Address(ctx context.Context, instanceID, elasticIPv4AddressPool, elasticIPv4Address string) (string, error)
+		CreateUserdata(ctx context.Context, in *CreateUserDataParams) (string, error)
+		CreateInstance(ctx context.Context, in *CreateInstanceParams) (CreatedInstance, error)
+		InstanceState(ctx context.Context, instanceID string) (string, error)
+		DeleteKeyPair(ctx context.Context, keyPairName string) error
+		DeleteSecurityGroup(ctx context.Context, securityGroupName string) error
+		DeleteInstance(ctx context.Context, instanceID string) error
+	}
+
+	DefaultClient struct {
+		Client
+		client *ec2.Client
+		prefix string
+	}
+
+	CreateUserDataParams struct {
+		ProjectToken     string
+		CoreInstanceName string
+	}
+
+	ElasticIPAddressParams struct {
+		Pool, Address string
+	}
+
+	CreateInstanceParams struct {
+		Region,
+		CoreInstanceName,
+		CoreVersion,
+		KeyPairName,
+		SecurityGroupName,
+		InstanceType,
+		SubnetID string
+		UserData        *CreateUserDataParams
+		PublicIPAddress *ElasticIPAddressParams
+	}
+
+	CreatedInstance struct {
+		CoreInstanceName string `json:"-"`
+		types.MetadataAWS
+	}
+)
+
+func (i *CreatedInstance) String() string {
+	return fmt.Sprintf("instance-id: %s, instance-type: %s, privateIPv4: %s", i.EC2InstanceID, i.EC2InstanceType, i.PrivateIPv4)
+}
+
+func New(ctx context.Context, prefix, region, credentials, profileFile, profileName string) (*DefaultClient, error) {
+	var opts []func(options *awsconfig.LoadOptions) error
+
+	if region != "" {
+		opts = append(opts, awsconfig.WithRegion(region))
+	}
+
+	if credentials != "" {
+		opts = append(opts, awsconfig.WithSharedCredentialsFiles([]string{credentials}))
+	}
+
+	if profileFile != "" {
+		opts = append(opts, awsconfig.WithSharedConfigFiles([]string{profileFile}))
+	}
+
+	if profileName != "" {
+		opts = append(opts, awsconfig.WithSharedConfigProfile(profileName))
+	}
+
+	cfg, err := awsconfig.LoadDefaultConfig(ctx, opts...)
+	if err != nil {
+		return nil, err
+	}
+
+	return &DefaultClient{
+		client: ec2.NewFromConfig(cfg),
+		prefix: prefix,
+	}, nil
+}
+
+func (c *DefaultClient) getTags(resourceType awstypes.ResourceType) []awstypes.TagSpecification {
+	return []awstypes.TagSpecification{
+		{
+			ResourceType: resourceType,
+			Tags: []awstypes.Tag{
+				{
+					Key:   aws.String(coreInstanceTag),
+					Value: aws.String(c.prefix),
+				},
+			},
+		},
+	}
+}
+
+func (c *DefaultClient) InstanceState(ctx context.Context, instanceID string) (string, error) {
+	describeInstanceStatus := &ec2.DescribeInstanceStatusInput{
+		IncludeAllInstances: aws.Bool(true),
+		InstanceIds:         []string{instanceID},
+	}
+
+	instanceStatus, err := c.client.DescribeInstanceStatus(ctx, describeInstanceStatus)
+	if err != nil {
+		return "", err
+	}
+
+	return string(instanceStatus.InstanceStatuses[0].InstanceState.Name), nil
```

Comment:
should we check before if this guy `instanceStatus.InstanceStatuses[0]`have any `len` before? cant we get a nil from here?